### PR TITLE
feat: add NPU device support for embedded-python model executor.

### PR DESCRIPTION
**Pull request by yingxudeng**
Enable --model_impl=python on NPU builds by:
- Registering xllm_ops torch library for NPU (npu_ops_library.cpp).
- Adding NpuPagedAttentionBackend using FIA (TND layout) for prefill/decode.
- Implementing pure-Python fused_qk_norm_rope fallback for non-CUDA devices.
- Patching torch_npu import and TORCH_LIBRARY conflicts in embedded interpreter.
- Routing attention backend and graph runner by device type in executor.

## Change Type

- [ ] Bug fix
- [ ] New feature
- [ ] Performance improvement
- [ ] Refactor
- [ ] Documentation
- [ ] Test
- [ ] Build or CI

## Pull Request Checklist

Thank you for contributing to xLLM. Before requesting review, please make sure the following items are complete.

### PR Title and Commit Messages

- [ ] The PR title and each commit message follow the xLLM commit format: `<type>: <subject>`.

> Allowed types: feat, bugfix, docs, test, refactor, chore, style, revert, perf, model, build, release.
> The subject should use clear English, start with a verb, include at least 4 words, and end with `.`.

### Pre-commit Checks

- [ ] I have installed `pre-commit` by running `pip install pre-commit` or an equivalent command.
- [ ] I have installed the hooks with `pre-commit install`.
- [ ] I have run `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

### Self Review

- [ ] I have self-reviewed the code according to `.agents/skills/code-review/references/custom-code-style.md`, especially code written or assisted by AI.
- [ ] I have rebased this PR onto the latest `main` branch.

### Build and Test Coverage

- [ ] Tests have been added or updated as needed.
- [ ] CUDA: `python setup.py build test` has passed on a CUDA machine.
- [ ] NPU: `python setup.py build test` has passed on an NPU machine.
- [ ] MLU: `python setup.py build test` has passed on an MLU machine.

### CODE REVIEW

**Comment by yingxudeng** [comment]
## GSM8K 精度测试报告

### 测试结果

| 指标               | 结果              |
| ------------------ | ----------------- |
| **GSM8K accuracy** | **0.78** (78/100) |
| 样本数             | 100               |
| max_tokens         | 4096              |
| temperature        | 0                 |

### 精度正确性证据

1. **GSM8K 得分符合预期**：Qwen3-4B thinking 模式 GSM8K 得分 0.78，与该模型公开 benchmark 水平一致。

2. **输出质量检查**：随机抽样检查多个 sample，推理链完整、答案正确：
   - Sample 1: 3×3×60 = 540 ✓
   - Sample 2: 2÷2 + 2 = 3 ✓
   - Sample 4: 8×$5 + 8×$3 = $64 ✓
   - Sample 5: (16-3-4)×$2 = $18 ✓

3. **无乱码/无截断**：全部 100 条输出无异常字符、无语义断裂，thinking 推理链和最终答案均完整输出。

4. **算子调用链验证**：Python 模式通过 `torch.ops.npu.npu_fused_infer_attention_score` 直接调用 FIA 算子，prefill 走 sparse_mode=3 causal mask，decode 走 sparse_mode=0 block_table paged attention，与 C++ 侧调用同一底层 kernel。

### 测试环境

- **模型**: Qwen3-4B (thinking mode)
- **硬件**: NPU (Ascend 910B)
- **容器**: dyx-xllm-cann9-main (CANN 9.0)
- **评测工具**: evalscope 1.4.2, GSM8K (4-shot)
- **启动参数**: `--model_impl python --enable_chunked_prefill false`

### 评测命令

```bash
# 启动服务
xllm --model /export/home/models/Qwen3-4B \
  --model_impl python \
  --enable_chunked_prefill false \
  --port 6010

# 评测
evalscope eval \
  --model Qwen3-4B \
  --api-url http://localhost:6010/v1 \
  --api-key EMPTY \
  --eval-type server \
  --datasets gsm8k \
  --eval-backend Native \
  --generation-config '{"max_tokens": 4096, "temperature": 0}' \
  --eval-batch-size 8 \
  --limit 100
```

File: xllm/python/model_executor/executor.py [hunk 1]
```diff
@@ -48,38 +78,37 @@ def __init__(
...
-            sliding_window=first_attention.sliding_window,
-            device=first_parameter.device,
-            dtype=first_parameter.dtype,
+        self.attention_backend = _create_attention_backend(
+            first_attention, device, first_parameter.dtype
         )
 
         execution_model = model.model
         self.eager_runner = EagerRunner(execution_model, self.attention_backend)
-        self.decode_cuda_graph_runner = None
```
Comment (yingxudeng) [inline xllm/python/model_executor/executor.py:69]: 分开

File: xllm/python/model_executor/executor.py [hunk 2]
```diff
@@ -48,38 +78,37 @@ def __init__(
...
+        self.decode_graph_runner = None
```
Comment (yingxudeng) [inline xllm/python/model_executor/executor.py]: 分开，不要统一 cuda 和 npu的 graph runner

File: xllm/core/kernels/npu/npu_ops_library.cpp [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+                                     const torch::Tensor& cos_sin_cache,
+                                     bool interleaved,
+                                     const torch::Tensor& position_ids) {
+  const int64_t num_tokens = qkv.size(0);
+  const int64_t q_size = num_heads_q * head_dim;
+  const int64_t k_size = num_heads_k * head_dim;
+
+  // Q and K are narrow (non-contiguous) views of QKV; copy to contiguous
+  // buffers for NPU kernels, process, then write back.
+  auto q_flat = qkv.narrow(1, 0, q_size).contiguous();
```
Comment (yingxudeng) [inline xllm/core/kernels/npu/npu_ops_library.cpp]: C++ 侧不要这种操作，放到 python 侧

File: xllm/python/model_executor/executor.py [hunk 4]
```diff
@@ -17,14 +17,44 @@
 import torch
 import torch.nn as nn
 
-from xllm.python.attention.backend import AttentionMetadata, KVCache
-from xllm.python.attention.flashinfer import FlashInferBackend
+from xllm.python.attention.backend import AttentionBackend, AttentionMetadata, KVCache
```
Comment (yingxudeng) [inline xllm/python/model_executor/executor.py:20]: 这事公共的，增加单测，分为 cuda 和 npu ，确保单测能过

File: xllm/python/model_executor/executor.py [hunk 5]
```diff
@@ -17,14 +17,44 @@
 import torch
 import torch.nn as nn
```
Comment (yingxudeng) [inline xllm/python/model_executor/executor.py:18]: foward_context 里面可以加类似 get device

File: xllm/python/ops/compute.py [hunk 6]
```diff
@@ -57,16 +56,72 @@ def fused_qk_norm_rope(
...
     )
 
 
+def _fused_qk_norm_rope_npu(
+    qkv: torch.Tensor,
+    num_heads_q: int, num_heads_k: int,
+    head_dim: int, eps: float,
+    q_weight: torch.Tensor, k_weight: torch.Tensor,
+    cos_sin_cache: torch.Tensor,
+    position_ids: torch.Tensor,
```
Comment (zhang-minchao) [inline xllm/python/ops/compute.py]: NPU平台多层用这个算子可能不适合，npu上每次都用position id查询没有那么高效

File: xllm/core/runtime/py_executor_impl.cpp [hunk 7]
```diff
@@ -111,6 +117,8 @@ PYBIND11_EMBEDDED_MODULE(xllm_runtime, m) {
                              &AttentionMetadataView::kv_cu_seq_lens)
       .def_property_readonly("kv_seq_lens_host",
                              &AttentionMetadataView::kv_seq_lens_host)
+      .def_property_readonly("block_table", &AttentionMetadataView::block_table)
+      .def_property_readonly("kv_seq_lens", &AttentionMetadataView::kv_seq_lens)
```
Comment (zhang-minchao) [inline xllm/core/runtime/py_executor_impl.cpp:121]: npu上fia算子不需要q_seq_lens吗？

File: xllm/python/attention/npu_paged_attention.py [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+        actual_seq_q = list(range(1, batch + 1))
+        actual_seq_kv = kv_seq_lens_host.tolist()
+
+        output, _ = torch.ops.npu.npu_fused_infer_attention_score(
+            q_3d, k_flat, v_flat,
+            pse_shift=None,
+            atten_mask=None,
+            actual_seq_lengths=actual_seq_q,
+            actual_seq_lengths_kv=actual_seq_kv,
+            block_table=block_table.to(torch.int32),
```
Comment (zhang-minchao) [inline xllm/python/attention/npu_paged_attention.py]: block table类型不能在转换，单个step只能一次

File: xllm/python/attention/npu_paged_attention.py [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+    ) -> torch.Tensor:
+        block_table = metadata.block_table
+        kv_seq_lens_host = metadata.kv_seq_lens_host
+
+        block_size = k_cache.size(1)
+        k_flat = k_cache.view(k_cache.size(0), block_size, -1)
+        v_flat = v_cache.view(v_cache.size(0), block_size, -1)
+
+        batch = kv_seq_lens_host.size(0)
+        actual_seq_q = list(range(1, batch + 1))
```
Comment (zhang-minchao) [inline xllm/python/attention/npu_paged_attention.py]: 在这构造actual_seq_q？

File: xllm/core/runtime/py_executor_impl.cpp (hunk 7 above)
Comment (yingxudeng) [reply xllm/core/runtime/py_executor_impl.cpp:121]: NPU FIA (TND layout) 的 actual_seq_lengths 参数语义是 cumulative sequence lengths，不是 per-sequence lengths。当前 Python 侧通过 q_cu_seq_lens[1:] 取累积值传入 FIA，已满足需求，无需额外暴露 q_seq_lens。

参照 C++ 路径（attention.cpp:102）传的也是 q_cu_seq_lens_host_vec（cumsum），以及 vllm-ascend 传的 query_start_loc_cpu[1:]（同为 cumulative）

**Comment by yingxudeng** [comment]
## Test Results

  ### Environment
  - Container: dyx-xllm-cann9-npu-python-graph
  - Device: Ascend 910 (single NPU)
  - Model: Qwen3-4B (model_impl=python)
  - Config: python_graph_backend=off, enable_prefix_cache=false, enable_chunked_prefill=false, enable_graph=false

  ### Unit Tests
  python setup.py build test --device npu
  100% tests passed, 0 tests failed out of 783
  (including NpuXllmOpsTest: 4/4 passed)

  ### End-to-End Accuracy (EvalScope, concurrent)
  evalscope eval --model Qwen3-4B --api-url http://localhost:9101/v1 \
    --datasets gsm8k --limit 10 --eval-batch-size 4

  +----------+---------+----------+--------+-----+-------+---------+
  | Model    | Dataset | Metric   | Subset | Num | Score | Cat.0   |
  +==========+=========+==========+========+=====+=======+=========+
  | Qwen3-4B | gsm8k   | mean_acc | main   |  10 |  0.90 | default |
  +----------+---------+----------+--------+-----+-------+---------+

  ### Manual Sanity Checks
  - "中国的首都是哪里？" → "北京。" (finish_reason: stop)
  - "Count 1 to 5" → "1 2 3 4 5" (finish_reason: stop)
  - Long generation (513 tokens): coherent, no garbled text (finish_reason: stop)

**Comment by yingxudeng** [comment]
## Test Results

### Environment
- Container: dyx-xllm-cann9-npu-python-graph
- Device: Ascend 910 (single NPU)
- Model: Qwen3-4B (model_impl=python)
- Config: python_graph_backend=off, enable_prefix_cache=false, enable_chunked_prefill=false, enable_graph=false

### Unit Tests
```bash
python setup.py build test --device npu
```
100% tests passed, 0 tests failed out of 783
(including NpuXllmOpsTest: 4/4 passed)

### End-to-End Accuracy (EvalScope, batch_size=4)
```bash
evalscope eval --model Qwen3-4B --api-url http://localhost:9101/v1 \
  --datasets gsm8k --limit 10 --eval-batch-size 4
```

| Model    | Dataset | Metric   | Subset | Num | Score | Cat.0   |
| -------- | ------- | -------- | ------ | --- | ----- | ------- |
| Qwen3-4B | gsm8k   | mean_acc | main   | 10  | 1.0   | default |

### Manual Sanity Checks
- "中国的首都是哪里？" → "北京。" (finish_reason: stop)
- "Count 1 to 5" → "1 2 3 4 5" (finish_reason: stop)
- Long generation (513 tokens): coherent, no garbled text (finish_reason: stop)

**Comment by yingxudeng** [comment]
## GPU 端到端精度验证

**环境：**
- 模型：Qwen3-4B (bf16)
- 硬件：NVIDIA H800 单卡
- 后端：`--model_impl=python --python_graph_backend=cudagraphs`
- 容器：`quay.io/jd_xllm/xllm-ai:xllm-dev-cuda-x86`
- flashinfer：0.6.14

**评测结果（evalscope, GSM8K 10-sample, 4-shot CoT, batch_size=4）：**

```
┌──────────┬─────────┬──────────┬────────┬─────┬───────┐
│ Model    │ Dataset │ Metric   │ Subset │ Num │ Score │
├──────────┼─────────┼──────────┼────────┼─────┼───────┤
│ Qwen3-4B │ gsm8k   │ mean_acc │ main   │  10 │  1.0  │
└──────────┴─────────┴──────────┴────────┴─────┴───────┘
```

**性能：** Avg latency 8.02s, throughput 152 tok/s (avg_in=662, avg_out=1220)

**结论：** Python graph executor (cudagraphs) 在 CUDA 设备上精度正确，GSM8K 4-shot 10条全部通过。

**Review by zhang-minchao — APPROVED** [review]

**Review by weizhehuang0827 — APPROVED** [review]
